src/core/Present.jl

```diff
@@ -120,94 +120,98 @@ end
 """
 generator_index(pres::Presentation, x::Symbol) = pres.generator_name_index[x].second
 
-# Presentation macro
-####################
-
-""" Define a presentation using a convenient syntax.
+""" Shorthand for contructing a term in the presentation
 """
-macro present(head, body)
-  name, pres = @match head begin
-    Expr(:call, name::Symbol, syntax_name::Symbol) =>
-      (name, :($(module_ref(:Presentation))($syntax_name)))
-    Expr(:(<:), name::Symbol, parent::Symbol) => (name, :(copy($parent)))
-    _ => throw(ParseError("Ill-formed presentation header $head"))
+function (pres::Presentation)(e::Symbol)
+  pres[e]
+end
+
+function (pres::Presentation)(e::Expr)
+  @match e begin
+    Expr(:call, term_constructor, args...) =>
+      invoke_term(pres.syntax, term_constructor, map(e -> pres(e), args)...)
   end
-  let_expr = Expr(:let, Expr(:block), translate_presentation(pres, body))
-  esc(Expr(:(=), name, let_expr))
 end
 
-""" Translate a presentation in the DSL to a block of Julia code.
+
+""" Create a new generator in a presentation of a given type
 """
-function translate_presentation(pres::Expr0, body::Expr)::Expr
-  @assert body.head == :block
-  code = Expr(:block)
-  append_expr!(code, :(_presentation = $pres))
-  for expr in strip_lines(body).args
-    append_expr!(code, translate_statement(expr))
-  end
-  append_expr!(code, :(_presentation))
-  code
+function make_generator(pres::Presentation, name::Symbol,
+              type::Symbol, type_args::Vector)
+  invoke_term(pres.syntax, type, name, map(e -> pres(e), type_args)...)
 end
 
-""" Translate a single statement in the presentation DSL to Julia code.
+""" Create and add a new generator
 """
-function translate_statement(expr::Expr)::Expr
-  @match expr begin
-    Expr(:(::), name::Symbol, type_expr) =>
-      translate_generator(name, type_expr)
-    Expr(:(::), Expr(:tuple, names...), type_expr) =>
-      Expr(:block, (translate_generator(name, type_expr) for name in names)...)
-    Expr(:(::), type_expr) =>
-      translate_generator(nothing, type_expr)
-    Expr(:(:=), name::Symbol, def_expr) =>
-      translate_definition(name, def_expr)
-    Expr(:call, :(==), lhs, rhs) => translate_equation(lhs, rhs)
-    _ => throw(ParseError("Ill-formed presentation statement $expr"))
+function add_generator!(pres::Presentation, name::Symbol,
+                   type::Symbol, type_args::Vector=[]; idem=true)
+  if !(name ∈ keys(pres.generator_name_index)) || !idem
+    add_generator!(pres, make_generator(pres,name,type,type_args))
   end
 end
 
-""" Translate declaration of a generator.
+""" Create and add multiple generators
 """
-function translate_generator(name::Union{Symbol,Nothing}, type_expr)::Expr
-  type_name, args = @match type_expr begin
-    name::Symbol => (name, [])
-    Expr(:call, name::Symbol, args...) => (name, args)
-    _ => throw(ParseError("Ill-formed type expression $type_expr"))
+function add_generators!(pres::Presentation, names::Vector,
```

Comment:
Do we want to overload `add_generator(s)!` with these extra methods? I feel like this is an implementation detail of the `@present` macro and should not be exposed to the user through these generic functions.